pennylane_lightning/src/simulator/Measures.hpp

```diff
@@ -309,6 +310,98 @@ class Measures {
         return expected_value_list;
     };
 
+    /**
+     * @brief Complete a single Metropolis-Hastings step.
+     *
+     * @param sv state vector
+     * @param tk User-defined functor for producing transitions
+     * between metropolis states.
+     * @param gen Random number generator.
+     * @param distrib Random number distribution.
+     * @param s1 initial state
+     */
+    size_t metropolis_step(const SVType &sv,
+                           std::unique_ptr<TransitionKernel<fp_t>> &tk,
+                           std::mt19937 &gen,
+                           std::uniform_real_distribution<fp_t> &distrib,
+                           size_t s1) {
+        auto s1_plog =
+            std::log((sv.getData()[s1] * std::conj(sv.getData()[s1])).real());
+
+        auto s1_qratio = tk->operator()(s1);
+
+        // transition kernel outputs these two
+        auto &s2 = s1_qratio.first;
+        auto &qratio = s1_qratio.second;
+
+        auto s2_plog =
+            std::log((sv.getData()[s2] * std::conj(sv.getData()[s2])).real());
+
+        auto alph = std::min<fp_t>(1., qratio * std::exp(s2_plog - s1_plog));
+        auto ran = distrib(gen);
+
+        if (ran < alph) {
+            return s2;
+        }
+        return s1;
+    }
+
+    /**
+     * @brief Generate samples using the Metropolis-Hastings method.
+     * Reference: Numerical Recipes, NetKet paper
+     *
+     * @param transition_kernel User-defined functor for producing transitions
+     * between metropolis states.
+     * @param num_burnin Number of Metropolis burn-in steps.
+     * @param num_samples The number of samples to generate.
+     * @return 1-D vector of samples in binary, each sample is
+     * separated by a stride equal to the number of qubits.
+     */
+    std::vector<size_t>
+    generate_samples_metropolis(const TransitionKernelType &transition_kernel,
```

Comment:
Is it possible to make the transition_kernel name the input here?
In this way, we could perform the transition_kernel selection inside this method instead of doing it in the bindings as it is now. I left a comment there too.

Reply:
Yep, I'd also like to see this.